pkg/jobs/jobs.go

```diff
@@ -0,0 +1,128 @@
+package jobs
+
+import (
+	"bytes"
+	"encoding/gob"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"time"
+)
+
+const (
+	// Queued state
+	Queued State = "queued"
+	// Running state
+	Running = "running"
+	// Errored state
+	Errored = "errored"
+)
+
+const (
+	// JSONEncoding is a JSON encoding message type
+	JSONEncoding = "json"
+	// GobEncoding is a Gob encoding message type
+	GobEncoding = "gob"
+)
+
+type (
+	// Queue interface is used to represent a asynchronous queue of jobs from
+	// which it is possible to enqueue and consume jobs.
+	Queue interface {
+		Enqueue(*Job) error
+		Consume() (bool, *Job, error)
+		Foreach(func(*Job) error) error
+		Len() int
+		Close()
+	}
+
+	// Broker interface is used to represent a job broker associated to a
+	// particular domain from which it is possible to create jobs that a pushed
+	// in the job system.
+	Broker interface {
+		Domain() string
+		PushJob(*JobRequest) (*Job, error)
+	}
+
+	// State represent the state of a job.
+	State string
+
+	// Message is a byte slice representing an encoded job message type.
+	Message struct {
+		Data []byte
+		Type string
+	}
+
+	// Job struct encapsulates all the parameters of a job.
+	Job struct {
+		ID         string    `json:"id"`
+		WorkerType string    `json:"worker_type"`
+		Message    *Message  `json:"-"`
+		State      State     `json:"state"`
+		QueuedAt   time.Time `json:"queued_at"`
+	}
+
+	// JobRequest struct is used to represent a new job request.
+	JobRequest struct {
+		WorkerType string
+		Message    *Message
+	}
+)
+
+// NewMessage returns a new Message encoded in the specified format.
+func NewMessage(enc string, data interface{}) (*Message, error) {
+	var b []byte
+	var err error
+	switch enc {
+	case JSONEncoding:
+		b, err = json.Marshal(data)
+	case GobEncoding:
+		buf := new(bytes.Buffer)
+		if err = gob.NewEncoder(buf).Encode(data); err != nil {
+			return nil, err
+		}
+		b = buf.Bytes()
+	default:
+		err = ErrUnknownMessageType
+	}
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type: enc,
+		Data: b,
+	}, nil
+}
+
+// Unmarshal can be used to unmarshal the encoded message value in the
+// specified interface's type.
+func (m *Message) Unmarshal(msg interface{}) error {
+	switch m.Type {
+	case JSONEncoding:
+		return json.NewDecoder(bytes.NewReader(m.Data)).Decode(msg)
+	case GobEncoding:
+		return gob.NewDecoder(bytes.NewReader(m.Data)).Decode(msg)
+	default:
+		return ErrUnknownMessageType
+	}
+}
+
+func makeQueueName(domain, workerType string) string {
+	return fmt.Sprintf("%s/%s", domain, workerType)
+}
+
+func makeID() string {
+	const (
+		idLen         = 16
+		letters       = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+		letterIdxMask = 1<<6 - 1 // 6 bits to represent a letter index
+	)
+	b := make([]byte, idLen)
+	for i := 0; i < idLen; {
+		if idx := int(rand.Int63() & letterIdxMask); idx < len(letters) {
```

Comment:
And shouldn't we run `rand.Seed` to avoid having the same ids when the cozy-stack is restarted?